system/include/SSE/xmmintrin.h

```diff
@@ -0,0 +1,709 @@
+_mm_store_ps(float *__p, __m128 __a)
+{
+  *(__m128 *)__p = __a;
+}
+// No NTA cache hint available.
+#define _mm_stream_ps _mm_store_ps
+
+#define _MM_HINT_T0 3
+#define _MM_HINT_T1 2
+#define _MM_HINT_T2 1
+#define _MM_HINT_NTA 0
+// No prefetch available, dummy it out.
+static __inline__ void __attribute__((__always_inline__))
+_mm_prefetch(const void *__p, int __i)
+{
+  ((void)__p);
+  ((void)__i);
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_sfence(void)
+{
+  __sync_synchronize(); // Emscripten/SharedArrayBuffer has only a full barrier instruction, which gives a stronger guarantee.
+}
+
+#define _MM_SHUFFLE(w, z, y, x) (((w) << 6) | ((z) << 4) | ((y) << 2) | (x))
+
+// This is defined as a macro because __builtin_shufflevector requires its
+// mask argument to be a compile-time constant.
+#define _mm_shuffle_ps(a, b, mask) \
+  ((__m128)__builtin_shufflevector((__f32x4)(a), (__f32x4)(b), \
+                                  (((mask) >> 0) & 0x3) + 0, \
+                                  (((mask) >> 2) & 0x3) + 0, \
+                                  (((mask) >> 4) & 0x3) + 4, \
+                                  (((mask) >> 6) & 0x3) + 4))
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storer_ps(float *__p, __m128 __a)
+{
+  _mm_store_ps(__p, _mm_shuffle_ps(__a, __a, _MM_SHUFFLE(0, 1, 2, 3)));
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ps1(float *__p, __m128 __a)
+{
+  _mm_store_ps(__p, _mm_shuffle_ps(__a, __a, _MM_SHUFFLE(0, 0, 0, 0)));
+}
+#define _mm_store1_ps _mm_store_ps1
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ss(float *__p, __m128 __a)
+{
+    struct __unaligned {
+      float __v;
+    } __attribute__((__packed__, __may_alias__));
+    ((struct __unaligned *)__p)->__v = ((__f32x4)__a)[0];
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storeu_ps(float *__p, __m128 __a)
+{
+  struct __unaligned {
+    __m128 __v;
+  } __attribute__((__packed__, __may_alias__));
+  ((struct __unaligned *)__p)->__v = __a;
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storeu_si16(void *__p, __m128i __a)
+{
+  struct __unaligned {
+    unsigned short __u;
+  } __attribute__((__packed__, __may_alias__));
+  ((struct __unaligned *)__p)->__u = ((__u16x8)__a)[0];
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storeu_si64(void *__p, __m128i __a)
+{
+  struct __unaligned {
+    unsigned long long __u;
+  } __attribute__((__packed__, __may_alias__));
+  ((struct __unaligned *)__p)->__u = ((__u64x2)__a)[0];
+}
+
+static __inline__ int __attribute__((__always_inline__))
+_mm_movemask_ps(__m128 __a)
+{
+  union {
+    __m128 __v;
+    unsigned int __x[4];
+  } __attribute__((__packed__, __may_alias__)) __p;
+  __p.__v = __a;
+  return (__p.__x[0] >> 31)
+    | ((__p.__x[1] >> 30) & 2)
+    | ((__p.__x[2] >> 29) & 4)
+    | ((__p.__x[3] >> 28) & 8);
```

Comment:
Have you experimented with the prototyped `bitmask` instruction for this one?

Reply:
That one does not seem to be available yet (https://github.com/WebAssembly/simd/pull/201). That one looks like it would directly implement movemask here, so would be pretty cool to have that.

Added a TODO in code to remember to look at that when it becomes available.